### PR TITLE
Various enhancements

### PR DESCRIPTION
**Pull request by tvolkert**
* Provide a more helpful (contextual) exception message when
  we fail to set a field while decoding a property.

* Relax the constraints that dictate whether we found a
  "default constructor" when loading a model class. The
  existing constraints didn't allow a class to be used both
  as a model and as a json_serializable target, since the
  latter requires the default constructor to provide named
  arguments for all the properties.

* Add generics support to `withTransaction()`, making the signature:
  `Future<T> withTransaction<T>(TransactionHandler<T>)`.
  This enables type safety in the returned future vis-a-vis
  the return type of the transaction callback.

### CODE REVIEW

File: lib/src/db/model_db_impl.dart [hunk 1]
```diff
@@ -238,6 +238,17 @@ class ModelDBImpl implements ModelDB {
...
 
+  /// Returns true if a  constructor invocation is valid even if the specified
+  /// [parameter] is omitted.
+  ///
+  /// This is  true for named parameters, optional parameters, and parameters
+  /// with a default value.
+  static bool _canBeOmitted(mirrors.ParameterMirror parameter) {
+    return parameter.isOptional ||
+        parameter.isNamed ||
+        parameter.hasDefaultValue;
```
Comment (dnfield) [inline lib/src/db/model_db_impl.dart]: I'm not sure if it makes sense to check this, but does it make a difference if the parameter is `@required`?  Does mirrors even give you access to that?

File: lib/src/db/model_db_impl.dart [hunk 2]
```diff
@@ -439,7 +450,12 @@ class _ModelDescription<T extends Model> {
           '${entity.key.elements.last.kind} (property name: $propertyName)');
     }
 
-    mirror.setField(mirrors.MirrorSystem.getSymbol(fieldName), value);
+    try {
+      mirror.setField(mirrors.MirrorSystem.getSymbol(fieldName), value);
+    } catch (error) {
```
Comment (jonasfj) [inline lib/src/db/model_db_impl.dart]: nit. Could we be more specific about what we are catching?

File: lib/src/db/model_db_impl.dart (hunk 1 above)
Comment (tvolkert) [reply lib/src/db/model_db_impl.dart]: Done (and test added).

File: lib/src/db/model_db_impl.dart (hunk 2 above)
Comment (tvolkert) [reply lib/src/db/model_db_impl.dart]: Done

**Pull request merged**